reg_apport_prototype.py

```diff
@@ -0,0 +1,164 @@
+#%% -*- coding: utf-8 -*-
+"""
+Created on Wed Nov 22 11:46:52 2023
+@author: zoring
+
+"""
+import numpy as np
+import pandas as pd
+from pandas.api.types import is_numeric_dtype
+
+import os
+#%% Loading ddata
+mydir = r"D:\data\res_dev\outputs\reg_apport"
+in_file = "outputs_df_before.pkl"
+out_file = "outputs_df_corrected.csv"
+
+#%% Parameters
+cols = {
+    "ref": "reference",
+    "ins": "instance",
+    "period": "period",
+    "form": "formtype",
+    "postcode": "postcodes_harmonised",
+    "percent": "602",
+    "product": "201",
+    "civdef": "200",
+}
+
+want_cals = [str(x) for x in range(202, 509)] 
+
+short_code = "0006"
+long_code = "0001"
+short_percent = 100.0
+
+
+# %%  Functions
+
+# Cleaning the short forms
+def run_short(df: pd.DataFrame) -> pd.DataFrame:
+    df_out = df.copy()
+    cond = (df_out[cols["form"]] == short_code)
+    df_out[cols["percent"]].mask(cond, other=short_percent, inplace=True)
+    return df_out
+
+# Counting unique non-blank codes
+def count_codes(df: pd.DataFrame, code: str = "postcode") -> pd.DataFrame:
+    # Calculates the number of unique non-empty codes
+
+    dfa = df.copy()
+    # Select columns that we need
+    cols_need = [cols["ref"], cols["period"], cols[code]]
+    dfa = dfa[cols_need]
+    dfa = dfa[dfa[cols[code]].str.len() > 0]
+    dfa.drop_duplicates(inplace=True)
+    dfb = dfa.groupby([cols["ref"], cols["period"]]).agg("count").reset_index()
+    dfb.rename({cols[code]: cols[code] + "_count"}, axis='columns', inplace=True)
+    df = df.merge(
+        dfb,
+        on = [cols["ref"], cols["period"]],
+        how="left")
+    return df
+
+# Computes the total numerical value across multiple sites and spreads it using 
+# the weights
+def value_to_sites(dfc : pd.DataFrame, vc: str) -> pd.DataFrame:
+    vcs = vc + "_s"
+    dfc[vcs] = dfc[vc].fillna(0)
+    dfc[vcs] = dfc.groupby([cols["ref"], cols["period"]])[vcs].transform("sum")
+    dfc[vcs] = dfc[vcs] * dfc["s_weight"]
+    dfc[vcs].replace(0, np.nan, inplace=True)
+    return dfc
+
+# Calculate weights
+def weights(dfc):
```

Comment:
@zorge69 please write a docstring for this function. And comment the process please.